local Charm = require("../../Charm")
local ReactCharm = require("../")
local renderHook = require("./helpers/renderHook")
local suite = require("../../../../test/suite")

local test = suite.test

test("tracks signal state", function()
	local getState, setState = Charm.signal(0)
	local hook = renderHook(function()
		return ReactCharm.useSignalState(getState)
	end)
	assert(hook.result == 0, "should return 0")
	setState(1)
	assert(hook.result == 1, "should return 1")
end)

test("accepts a selector", function()
	local getState, setState = Charm.signal(0)
	local hook = renderHook(function()
		return ReactCharm.useSignalState(function()
			return 2 * getState()
		end)
	end)
	assert(hook.result == 0, "should return 0")
	setState(1)
	assert(hook.result == 2, "should return 2")
end)

test("tracks dependencies", function()
	local getState, setState = Charm.signal(0)
	local hook = renderHook(function()
		return ReactCharm.useSignalState(getState, { getState })
	end)
	assert(hook.result == 0, "should return 0")
	setState(1)
	assert(hook.result == 1, "should return 1")

	getState, setState = Charm.signal(0)
	hook.rerender()
	assert(hook.result == 0, "should return 0")
	setState(1)
	assert(hook.result == 1, "should return 1")
end)

return {}
