--!optimize 1

local Charm = require("../src")
local test = require("../../../tests/test")

test("should disallow yielding", function()
	local success = pcall(function()
		Charm.effectScope(function()
			coroutine.yield()
		end)
	end)
	assert(not success, "should error on yield")
end)

test.describe("traceback", function()
	local DEFAULT_ERROR = "Test error"

	local function fail(message: string?)
		error(message or DEFAULT_ERROR, 0)
	end

	local function indirectError(message: string?)
		fail(message)
	end

	local function assertStackTrace(err: unknown, expectation: string?)
		expectation = expectation or DEFAULT_ERROR
		assert(type(err) == "string", "Expected error to be a string")
		assert(string.find(err, expectation), `Expected error message to be "{expectation}", got: {err}`)
		assert(string.find(err, "fail"), `Expected stack trace to contain fail, got: {err}`)
		assert(string.find(err, "indirectError"), `Expected stack trace to contain indirectError, got: {err}`)
	end

	test("signal update", function()
		local _, err = pcall(function()
			local _, setSrc = Charm.signal(0)
			setSrc(function()
				indirectError()
				return 1
			end)
		end)
		assertStackTrace(err)
	end)

	test("atom update", function()
		local _, err = pcall(function()
			local src = Charm.atom(0)
			src(function()
				indirectError()
				return 1
			end)
		end)
		assertStackTrace(err)
	end)

	test("computed", function()
		local _, err = pcall(function()
			local src = Charm.computed(function()
				indirectError()
			end)
			src()
		end)
		assertStackTrace(err)
	end)

	test("effect init", function()
		local _, err = pcall(function()
			Charm.effect(function()
				indirectError()
			end)
		end)
		assertStackTrace(err)
	end)

	test("effect run", function()
		local _, err = pcall(function()
			local src, setSrc = Charm.signal(0)
			Charm.effect(function()
				if src() == 1 then
					indirectError()
				end
			end)
			setSrc(1)
		end)
		assertStackTrace(err)
	end)

	test("effect cleanups", function()
		local _, err = pcall(function()
			local dispose = Charm.effect(function()
				Charm.onCleanup(function()
					indirectError("onCleanup")
				end)
				return function()
					indirectError("returned cleanup")
				end
			end)
			dispose()
		end)
		assertStackTrace(err, "onCleanup")
		assertStackTrace(err, "returned cleanup")
	end)

	test("trigger", function()
		local _, err = pcall(function()
			Charm.trigger(function()
				indirectError()
			end)
		end)
		assertStackTrace(err)
	end)

	test("untracked", function()
		local _, err = pcall(function()
			Charm.effectScope(function()
				Charm.untracked(function()
					indirectError()
				end)
			end)
		end)
		assertStackTrace(err)
	end)

	test("batch", function()
		local _, err = pcall(function()
			Charm.batch(function()
				indirectError()
			end)
		end)
		assertStackTrace(err)
	end)

	test("subscribe getter", function()
		local _, err = pcall(function()
			Charm.subscribe(function()
				indirectError()
			end, function() end)
		end)
		assertStackTrace(err)
	end)

	test("subscribe callback", function()
		local _, err = pcall(function()
			local src, setSrc = Charm.signal(0)
			Charm.subscribe(src, function()
				indirectError()
			end)
			setSrc(1)
		end)
		assertStackTrace(err)
	end)

	test("listen getter", function()
		local _, err = pcall(function()
			Charm.listen(function()
				indirectError()
			end, function() end)
		end)
		assertStackTrace(err)
	end)

	test("listen callback", function()
		local _, err = pcall(function()
			local src, setSrc = Charm.signal(0)
			Charm.listen(src, function()
				indirectError()
			end)
			setSrc(1)
		end)
		assertStackTrace(err)
	end)

	test("mapped getter", function()
		local _, err = pcall(function()
			local src = Charm.mapped(function()
				indirectError()
				return { 1 }
			end, function()
				return 1
			end)
			src()
		end)
		assertStackTrace(err)
	end)

	test("mapped transform", function()
		local _, err = pcall(function()
			local src = Charm.mapped(function()
				return { 1 }
			end, function()
				indirectError()
				return 1
			end)
			src()
		end)
		assertStackTrace(err)
	end)
end)

return {}
