local atom = require(script.Parent.Parent.atom)
local store = require(script.Parent.Parent.store)

return function()
	it("gracefully handles yielding", function()
		expect(function()
			store.capture(function()
				task.wait(10)
				return 1
			end)
		end).to.throw()
	end)

	it("captures atoms from a function", function()
		local a, b, c = atom(1), atom(1), atom(2)
		local deps, result = store.capture(function()
			return a() + b()
		end)
		expect(deps[a]).to.be.ok()
		expect(deps[b]).to.be.ok()
		expect(deps[c]).to.never.be.ok()
		expect(result).to.equal(2)
	end)

	it("captures an atom by itself", function()
		local a = atom(1)
		local deps, result = store.capture(a)
		expect(deps[a]).to.be.ok()
		expect(result).to.equal(1)
	end)
end
