local suite = require("@test/suite")

suite.describe("signals", function()
	suite.describe("computed", function()
		require("./computed.spec")
	end)
	suite.describe("effect", function()
		require("./effect.spec")
	end)
	suite.describe("effectScope", function()
		require("./effectScope.spec")
	end)
	suite.describe("issue_48", function()
		require("./issue_48.spec")
	end)
	suite.describe("gc", function()
		require("./gc.spec")
	end)
	suite.describe("topology", function()
		require("./topology.spec")
	end)
	suite.describe("untrack", function()
		require("./untrack.spec")
	end)
end)

return {}
