local suite = require("../../../../test/suite")

suite.describe("computed", function()
	require("./computed.spec")
end)
suite.describe("effect", function()
	require("./effect.spec")
end)
suite.describe("effectScope", function()
	require("./effectScope.spec")
end)
suite.describe("gc", function()
	require("./gc.spec")
end)
suite.describe("listen", function()
	require("./listen.spec")
end)
suite.describe("mapped", function()
	require("./mapped.spec")
end)
suite.describe("observe", function()
	require("./observe.spec")
end)
suite.describe("peek", function()
	require("./peek.spec")
end)
suite.describe("signal", function()
	require("./signal.spec")
end)
suite.describe("subscribe", function()
	require("./subscribe.spec")
end)
suite.describe("topology", function()
	require("./topology.spec")
end)
suite.describe("trigger", function()
	require("./trigger.spec")
end)
suite.describe("untracked", function()
	require("./untracked.spec")
end)

return {}
