local suite = require("../../../../test/suite")

suite.describe("core", function()
	suite.describe("computed", function()
		require("./core/computed.spec")
	end)
	suite.describe("effect", function()
		require("./core/effect.spec")
	end)
	suite.describe("effectScope", function()
		require("./core/effectScope.spec")
	end)
	suite.describe("issue_48", function()
		require("./core/issue_48.spec")
	end)
	suite.describe("gc", function()
		require("./core/gc.spec")
	end)
	suite.describe("topology", function()
		require("./core/topology.spec")
	end)
	suite.describe("untrack", function()
		require("./core/untrack.spec")
	end)
end)

suite.describe("listen", function()
	require("./listen.spec")
end)
suite.describe("mapped", function()
	require("./mapped.spec")
end)
suite.describe("observe", function()
	require("./observe.spec")
end)
suite.describe("subscribe", function()
	require("./subscribe.spec")
end)
suite.describe("untracked", function()
	require("./untracked.spec")
end)

return {}
